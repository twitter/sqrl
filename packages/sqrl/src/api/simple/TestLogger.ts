import { AbstractLogger } from "../../util/Logger";
import { LogProperties } from "../log";
import * as util from "util";

export class TestLogger extends AbstractLogger {
  readonly messages: {
    msg: string;
    level: string;
  }[] = [];

  countErrors() {
    let errorCount = 0;
    for (const { level } of this.messages) {
      if (level === "trace" || level === "debug" || level === "info") {
        continue;
      }
      errorCount += 1;
    }
    return errorCount;
  }

  popLatest(props: { level: string }) {
    const { level } = props;
    for (let idx = this.messages.length - 1; idx >= 0; idx--) {
      if (this.messages[idx].level === level) {
        return this.messages.splice(idx, 1)[0];
      }
    }
    throw new Error("No log line found");
  }

  log(level: string, props: LogProperties, format: string, ...param: any[]) {
    const message = util.format(format, ...param);
    // tslint:disable-next-line
    this.messages.push({
      msg: message,
      level
    });
    // tslint:disable-next-line:no-console
    console.error(message);
  }
}
