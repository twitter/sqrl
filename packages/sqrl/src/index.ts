/**
 * Copyright 2018 Twitter, Inc.
 * Licensed under the Apache License, Version 2.0
 * http://www.apache.org/licenses/LICENSE-2.0
 */
export * from "./api/ast";
export * from "./api/ctx";
export * from "./api/execute";
export * from "./api/expr";
export * from "./api/filesystem";
export * from "./api/log";
export * from "./api/parse";
