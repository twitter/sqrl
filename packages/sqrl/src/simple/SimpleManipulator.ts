/**
 * Copyright 2018 Twitter, Inc.
 * Licensed under the Apache License, Version 2.0
 * http://www.apache.org/licenses/LICENSE-2.0
 */
import { Manipulator, ManipulatorCallback } from "../platform/Manipulator";
import { kafkaBufferHumanJson } from "../jslib/kafkaBufferHumanJson";
import mapObject from "../jslib/mapObject";
import SqrlKey from "../object/SqrlKey";
import { WhenCause, FiredRule } from "../function/WhenFunctions";
import { FeatureMap } from "..";

export class SimpleManipulator extends Manipulator {
  public sqrlKeys: Set<string> = new Set();
  public kafkaOutput: {
    [topic: string]: Buffer[];
  } = {};

  public blocked: boolean = false;
  public whitelisted: boolean = false;
  public blockedRules: FiredRule[] = [];
  public whitelistedRules: FiredRule[] = [];

  public loggedFeatures: FeatureMap = {};
  public logged: string[] = [];

  addKafka(topic: string, message: Buffer) {
    this.kafkaOutput[topic] = this.kafkaOutput[topic] || [];
    this.kafkaOutput[topic].push(message);
  }

  setBlocked(context: WhenCause | null) {
    this.blocked = true;
    if (context) {
      this.blockedRules.push(...context.firedRules);
    }
  }
  setWhitelisted(context: WhenCause | null) {
    this.whitelisted = true;
    if (context) {
      this.whitelistedRules.push(...context.firedRules);
    }
  }
  wasBlocked() {
    return this.blocked && !this.whitelisted;
  }

  getCurrentHumanOutput() {
    return {
      kafka: mapObject(this.kafkaOutput, (messages, topic) => {
        return messages.map(msg => {
          return kafkaBufferHumanJson(topic, msg);
        });
      })
    };
  }

  trackSqrlKey(key: SqrlKey): void {
    this.sqrlKeys.add(key.getDebugString());
  }

  log(message: string) {
    this.logged.push(message);
  }
  logFeature(name: string, value: any) {
    this.loggedFeatures[name] = value;
  }
  logError(props: any): void {
    /* do nothing for now */
  }

  async mutate(ctx): Promise<void> {
    await Promise.all(this.callbacks.map(cb => cb(ctx)));
  }
  private callbacks: ManipulatorCallback[] = [];
  addCallback(cb: ManipulatorCallback) {
    this.callbacks.push(cb);
  }
}
