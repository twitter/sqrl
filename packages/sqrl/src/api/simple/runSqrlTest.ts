import { SimpleManipulator } from "./SimpleManipulator";
import { createSimpleContext } from "../ctx";
import { SqrlTest } from "../../testing/SqrlTest";
import { LocalFilesystem, Filesystem } from "../filesystem";
import {
  buildTestFunctionRegistry,
  buildTestServices
} from "../../testing/runSqrlTest";
import { invariant } from "sqrl-common";
import { FunctionServices } from "../../function/registerAllFunctions";
import { FunctionRegistry, Execution } from "../execute";
import * as path from "path";
import { Logger } from "../log";

export async function runSqrlTest(
  sqrl: string,
  options: {
    functionRegistry?: FunctionRegistry;
    services?: FunctionServices;
    logger?: Logger;
    filesystem?: Filesystem;
    librarySqrl?: string;
  } = {}
): Promise<{
  codedErrors: Error[];
  lastState: Execution;
  lastManipulator: SimpleManipulator;
}> {
  let functionRegistry: FunctionRegistry;

  let services: FunctionServices = {};
  if (options.functionRegistry) {
    invariant(
      !options.services,
      ".services not compatible with .functionRegistry"
    );
    functionRegistry = options.functionRegistry;
  } else {
    services = options.services || (await buildTestServices());
    functionRegistry = await buildTestFunctionRegistry({ services });
  }

  const filesystem =
    options.filesystem || new LocalFilesystem(path.join(__dirname, ".."));

  const test = new SqrlTest(functionRegistry._wrapped, {
    manipulatorFactory: () => new SimpleManipulator(),
    filesystem
  });
  const ctx = createSimpleContext(options.logger);
  if (options.librarySqrl) {
    await test.run(ctx, options.librarySqrl);
  }
  const rv = await test.run(ctx, sqrl);

  if (services.assert && services.assert.throwFirstError) {
    services.assert.throwFirstError();
  }

  return {
    ...rv,
    lastManipulator: rv.lastState.manipulator as SimpleManipulator
  };
}
