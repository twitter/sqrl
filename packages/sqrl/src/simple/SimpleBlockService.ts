/**
 * Copyright 2018 Twitter, Inc.
 * Licensed under the Apache License, Version 2.0
 * http://www.apache.org/licenses/LICENSE-2.0
 */
import { BlockService } from "../function/BlockFunctions";
import { Manipulator } from "../platform/Manipulator";
import { SimpleManipulator } from "../simple/SimpleManipulator";
import { WhenCause } from "../function/WhenFunctions";

export class SimpleBlockService implements BlockService {
  block(manipulator: Manipulator, cause: WhenCause | null) {
    if (!(manipulator instanceof SimpleManipulator)) {
      throw new Error("Expected SimpleManipulator for SimpleBlockService");
    }
    manipulator.setBlocked(cause);
  }
  whitelist(manipulator: Manipulator, cause: WhenCause | null) {
    if (!(manipulator instanceof SimpleManipulator)) {
      throw new Error("Expected SimpleManipulator for SimpleBlockService");
    }
    manipulator.setWhitelisted(cause);
  }
  wasBlocked(manipulator: Manipulator): boolean {
    if (!(manipulator instanceof SimpleManipulator)) {
      throw new Error("Expected SimpleManipulator for SimpleBlockService");
    }
    return manipulator.wasBlocked();
  }
}
