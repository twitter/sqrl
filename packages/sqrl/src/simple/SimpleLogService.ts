/**
 * Copyright 2018 Twitter, Inc.
 * Licensed under the Apache License, Version 2.0
 * http://www.apache.org/licenses/LICENSE-2.0
 */
import { Manipulator } from "../api/Manipulator";
import { SimpleManipulator } from "../simple/SimpleManipulator";
import { LogService } from "../function/LogFunctions";

export class SimpleLogService implements LogService {
  log(manipulator: Manipulator, message: string) {
    if (!(manipulator instanceof SimpleManipulator)) {
      throw new Error("Expected SimpleManipulator for SimpleBlockService");
    }
    manipulator.log(message);
  }
  logFeature(manipulator: Manipulator, name: string, value: any) {
    if (!(manipulator instanceof SimpleManipulator)) {
      throw new Error("Expected SimpleManipulator for SimpleBlockService");
    }
    manipulator.logFeature(name, value);
  }
}
