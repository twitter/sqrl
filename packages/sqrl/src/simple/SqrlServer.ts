/**
 * Copyright 2018 Twitter, Inc.
 * Licensed under the Apache License, Version 2.0
 * http://www.apache.org/licenses/LICENSE-2.0
 */
import { SimpleManipulator } from "./SimpleManipulator";
import { SqrlExecutionState } from "../execute/SqrlExecutionState";
import * as bluebird from "bluebird";
import * as micro from "micro";
import * as microQuery from "micro-query";
import { dispatch } from "micro-route";
import * as zipObject from "lodash.zipobject";
import { FeatureMap } from "../feature/FeatureTypes";
import { IncomingMessage, ServerResponse } from "http";
import SqrlExecutable from "../execute/SqrlExecutable";
import { FiredRule } from "../function/WhenFunctions";
import { Context } from "../api/ctx";

function userInvariant(cond, message) {
  if (!cond) {
    throw micro.createError(400, message);
  }
}

interface ApiRulesResponse {
  [ruleName: string]: {
    reason: string;
  };
}

interface ApiResponse {
  /** Overall block/allow verdict for the action */
  allow: boolean;

  /** Details about why the verdict was reached */
  verdict: {
    blockRules: string[];
    whitelistRules: string[];
  };

  rules: ApiRulesResponse;

  /** Returns features requested via the API */
  features?: FeatureMap;
}

async function run(
  ctx: Context,
  labeler: SqrlExecutable,
  req: IncomingMessage,
  res: ServerResponse
) {
  const query = microQuery(req);
  const featureTimeoutMs = parseInt(query.timeoutMs || "1000", 10);
  userInvariant(!isNaN(featureTimeoutMs), "timeoutMs was not a valid integer");

  const inputs = await micro.json(req, { limit: "128mb" });

  const manipulator = new SimpleManipulator();
  const execution: SqrlExecutionState = await labeler.startExecution(ctx, {
    manipulator,
    inputs,
    featureTimeoutMs
  });

  await execution.fetchBasicByName("SqrlExecutionComplete");

  const rules: ApiRulesResponse = {};
  function ruleToName(rule: FiredRule) {
    rules[rule.name] = {
      reason: rule.reason
    };
    return rule.name;
  }

  const rv: ApiResponse = {
    allow: !manipulator.wasBlocked(),
    verdict: {
      blockRules: manipulator.blockedRules.map(ruleToName) || [],
      whitelistRules: manipulator.whitelistedRules.map(ruleToName) || []
    },
    rules
  };

  if (query.features) {
    const featureNames = query.features.split(",");
    try {
      const values = await bluebird.map(featureNames, featureName =>
        execution.fetchBasicByName(featureName)
      );
      rv.features = zipObject(featureNames, values);
    } catch (e) {
      throw micro.createError(500, e.message, e);
    }
  }

  await manipulator.mutate(ctx);

  res.setHeader("Content-Type", "application/json");

  if (query.hasOwnProperty("pretty")) {
    return JSON.stringify(rv, undefined, 2) + "\n";
  } else {
    return JSON.stringify(rv);
  }
}

async function deleteRoute(
  ctx: Context,
  req: IncomingMessage,
  res: ServerResponse
) {
  // @TODO: Delete node
  throw micro.createError(500, "Not implemented\n");
}

export function createSqrlServer(ctx, labeler) {
  const router = dispatch()
    .dispatch("/run", ["POST"], (req, res) => run(ctx, labeler, req, res))
    .dispatch("/delete", ["POST"], (req, res) => deleteRoute(ctx, req, res))
    .otherwise(async (req, res) => {
      throw micro.createError(404, "Route not found\n");
    });

  return micro(router);
}
