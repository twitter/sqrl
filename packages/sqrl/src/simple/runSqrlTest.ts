/**
 * Copyright 2019 Twitter, Inc.
 * Licensed under the Apache License, Version 2.0
 * http://www.apache.org/licenses/LICENSE-2.0
 */
import { SimpleManipulator } from "./SimpleManipulator";
import { createSimpleContext } from "../api/ctx";
import { SqrlTest } from "../testing/SqrlTest";
import { LocalFilesystem, Filesystem } from "../api/filesystem";
import { buildTestFunctionRegistry } from "../testing/runSqrlTest";
import { FunctionRegistry, Execution } from "../api/execute";
import * as path from "path";
import { Logger } from "../api/log";
import { Config } from "../api/config";

export async function runSqrlTest(
  sqrl: string,
  options: {
    config?: Config;
    functionRegistry?: FunctionRegistry;
    logger?: Logger;
    filesystem?: Filesystem;
    librarySqrl?: string;
    register?: (instance: FunctionRegistry) => Promise<void>;
  } = {}
): Promise<{
  codedErrors: Error[];
  executions: Execution[];
  lastExecution: Execution;
  lastManipulator: SimpleManipulator;
}> {
  let functionRegistry: FunctionRegistry;

  if (options.functionRegistry) {
    functionRegistry = options.functionRegistry;
  } else {
    functionRegistry = await buildTestFunctionRegistry({
      config: options.config
    });
  }

  if (options.register) {
    await options.register(functionRegistry);
  }

  const filesystem =
    options.filesystem || new LocalFilesystem(path.join(__dirname, ".."));

  const test = new SqrlTest(functionRegistry._functionRegistry, {
    manipulatorFactory: () => new SimpleManipulator(),
    filesystem
  });
  const ctx = createSimpleContext(options.logger);
  if (options.librarySqrl) {
    await test.run(ctx, options.librarySqrl);
  }
  const rv = await test.run(ctx, sqrl);

  const lastExecution = rv.executions[rv.executions.length - 1];
  const lastManipulator = lastExecution.manipulator as SimpleManipulator;
  lastManipulator.throwFirstError();

  return {
    ...rv,
    lastExecution,
    lastManipulator
  };
}
