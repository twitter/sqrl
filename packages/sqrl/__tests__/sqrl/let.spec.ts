/**
 * Copyright 2018 Twitter, Inc.
 * Licensed under the Apache License, Version 2.0
 * http://www.apache.org/licenses/LICENSE-2.0
 */
import {
  runCompile,
  compileToExecution,
  fetchExecutableFeature
} from "../helpers/runCompile";
import { buildTestInstance } from "../../src/testing/runSqrlTest";
import { executableFromFilesystem, VirtualFilesystem } from "../../src";
import { runSqrlTest } from "../../src/simple/runSqrlTest";

test("supports basic statements", async () => {
  await expect(
    runCompile(`
LET A := FeatureOne + 1;
  `)
  ).rejects.toThrowError(/Feature was not defined: FeatureOne/);

  await expect(
    runCompile(`
LET FeatureOne := 1;
LET A := FeatureOne + 1;
LET A := FeatureOne + 2;
  `)
  ).rejects.toMatchSnapshot("multiple-definitions");

  const { execution } = await compileToExecution(`
LET FeatureOne := 1;
LET A := FeatureOne + 1;
_resetReplaceableFeatures();
LET A := FeatureOne + 2;
  `);

  await expect(execution.fetchBasicByName("A")).resolves.toEqual(3);
});

test("supports where clauses", async () => {
  const { executable } = await runCompile(`
  LET Select := input();
  LET A := 1 WHERE Select="one";
  LET A := 2 WHERE Select="two";
    `);

  await expect(
    fetchExecutableFeature(executable, "A", {
      inputs: {
        Select: "one"
      }
    })
  ).resolves.toEqual(1);
  await expect(
    fetchExecutableFeature(executable, "A", {
      inputs: {
        Select: "two"
      }
    })
  ).resolves.toEqual(2);

  await runSqrlTest(`
  LET TestActionName := '';
  LET Ip := '1.2.3.4' WHERE TestActionName = 'login';
  LET Ip := '5.6.7.8' WHERE TestActionName = 'signup';
  ASSERT Ip IS NULL;

  LET TestActionName := 'login';
  ASSERT Ip = '1.2.3.4';

  LET TestActionName := 'boom';
  LET IsPing := TestActionName = 'ping';
  LET IsPong := TestActionName = 'pong';
  LET Ball := 'red' WHERE IsPing or IsPong;
  LET Ball := 'blue' WHERE not (IsPing or IsPong);
  ASSERT Ball = 'blue';
  `);
});

test("supports where default values", async () => {
  await runSqrlTest(`
    LET TestActionName := '';
    LET Ip := 'loginIp' WHERE TestActionName = 'login';
    LET Ip := 'signupIp' WHERE TestActionName = 'signup';
    LET Ip := 'defaultIp' DEFAULT;
    ASSERT Ip = 'defaultIp';

    LET TestActionName := 'login';
    ASSERT Ip = 'loginIp';

    LET TestActionName := 'boom';
    ASSERT Ip = 'defaultIp';`);

  await expect(
    executableFromFilesystem(
      await buildTestInstance(),
      new VirtualFilesystem({
        "main.sqrl": `
    LET Act := "b";
    INCLUDE "a.sqrl" WHERE Act="a";
    ASSERT Act="c";
  `,
        "a.sqrl": `
  LET Something := "x" DEFAULT;
  `
      })
    )
  ).rejects.toThrow(
    /LET statement with DEFAULT is not valid in file included with WHERE/
  );

  const executable = await executableFromFilesystem(
    await buildTestInstance(),
    new VirtualFilesystem({
      "main.sqrl": `
    LET Choose := input();
    INCLUDE "default.sqrl";
    INCLUDE "red.sqrl" WHERE Choose="red";
    INCLUDE "blue.sqrl" WHERE Choose="blue";
  `,
      "default.sqrl": 'LET Color := "none" DEFAULT;',
      "blue.sqrl": 'LET Color := "cloudy blue";',
      "red.sqrl": 'LET Color := "deep dark red";'
    })
  );

  await expect(
    fetchExecutableFeature(executable._wrapped, "Color", {
      inputs: {
        Choose: "pink"
      }
    })
  ).resolves.toEqual("none");
  await expect(
    fetchExecutableFeature(executable._wrapped, "Color", {
      inputs: {
        Choose: "red"
      }
    })
  ).resolves.toEqual("deep dark red");
  await expect(
    fetchExecutableFeature(executable._wrapped, "Color", {
      inputs: {
        Choose: "blue"
      }
    })
  ).resolves.toEqual("cloudy blue");
});
