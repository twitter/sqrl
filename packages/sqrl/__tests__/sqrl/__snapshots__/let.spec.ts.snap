// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`supports basic statements: multiple-definitions 1`] = `
[Error: Multiple definitions of A
:3  LET A := FeatureOne + 1;]
`;
